src/lib/gossip_net/libp2p.ml

```diff
@@ -482,26 +483,20 @@ module Make (Rpc_intf : Mina_base.Rpc_intf.Rpc_interface_intf) :
               let logger = config.logger in
               [%log trace] ~metadata:[] "Successfully restarted libp2p" ;
               don't_wait_for (Strict_pipe.transfer m message_writer ~f:Fn.id))
-        and on_unexpected_termination () =
-          on_libp2p_create
-            (create_libp2p config rpc_handlers first_peer_ivar
-               high_connectivity_ivar ~added_seeds ~pids
-               ~on_unexpected_termination) ;
-          Deferred.unit
-        in
-        let res =
-          create_libp2p config rpc_handlers first_peer_ivar
-            high_connectivity_ivar ~added_seeds ~pids ~on_unexpected_termination
-        in
-        on_libp2p_create res ;
+        and start_libp2p () =
+          let libp2p =
+            create_libp2p config rpc_handlers first_peer_ivar
+              high_connectivity_ivar ~added_seeds ~pids
+              ~on_unexpected_termination:restart_libp2p
+          in
+          on_libp2p_create libp2p ; Deferred.ignore_m libp2p
+        and restart_libp2p () = don't_wait_for (start_libp2p ()) in
         don't_wait_for
           (Strict_pipe.Reader.iter restarts_r ~f:(fun () ->
                let%bind n = !net2_ref in
                let%bind () = Mina_net2.shutdown n in
-               let%bind () = on_unexpected_termination () in
-               !net2_ref >>| ignore)) ;
-        let%map _ = res in
-        ()
+               restart_libp2p () ; !net2_ref >>| ignore)) ;
```

Comment:
It's equivalent to the previous code, but a little easier to reason about (I think). The goal here is: "start the libp2p helper, then wait for the net2 instance to be assigned". In the previous code, we had to `bind` on `on_unexpected_termination ()` because its result was in the `Deferred.t` monad. In my refactor, I removed that monad nesting (it was unnecessary).

Reply:
The trailing `start_libp2p` call replaced the following equivalent code:

```ocaml
let res =
  create_libp2p config rpc_handlers first_peer_ivar
    high_connectivity_ivar ~added_seeds ~pids ~on_unexpected_termination
in
on_libp2p_create res ;
let%map _ = res in
()
```